ocaml-lsp-server/src/dune.ml

```diff
@@ -1,46 +1,159 @@
+  let pid =
+    let args = Array.of_list [ bin; "rpc"; "init" ] in
+    (* TODO collect stderr and report it properly *)
+    Unix.create_process bin args stdin_i stdout_o Unix.stderr
+    |> Stdune.Pid.of_int
+  in
+  let chan = Chan.create t.scheduler stdout_i stdin_o in
+  Fiber.fork_and_join_unit
+    (fun () ->
       let open Fiber.O in
-      let pid =
-        let args = Array.of_list [ bin; "rpc"; "--help=plain" ] in
-        Unix.create_process bin args stdin stdout stderr |> Stdune.Pid.of_int
-      in
-      let* status = Scheduler.wait_for_process t.scheduler pid in
-      let state =
-        match status with
-        (* TODO actually turn on *)
-        | Unix.WEXITED 0 -> Running
-        | _ -> Out_of_date
-      in
-      let+ () = Fiber.Ivar.fill t.state state in
-      state)
+      let* res = Scheduler.wait_for_process t.scheduler pid in
+      match res with
+      | Unix.WEXITED 0 -> Fiber.return ()
+      | _ ->
+        (* TODO inform client instead *)
+        failwith "Unexpected termination from dune")
+    (fun () ->
+      Client.connect_persistent chan
+        ~on_connected:(fun () t ->
+          let sub = Drpc.Subscribe.Diagnostics in
+          Client.notification t Drpc.Notification.subscribe sub)
+        ~on_connect:(fun () ->
+          let init =
+            Drpc.Initialize.create ~id:(Drpc.Id.make (Atom "ocamllsp"))
+          in
+          let diagnostic evs =
+            List.iter evs ~f:(fun (ev : Drpc.Diagnostic.Event.t) ->
+                let id =
+                  Poly.hash
+                    (match ev with
+                    | Add x -> x
+                    | Remove x -> x)
+                in
+                match ev with
+                | Remove _ -> Diagnostics.remove t.diagnostics (`Dune id)
+                | Add d ->
+                  (* TODO modify dune to return this reliably *)
+                  let uri : Uri.t = assert false in
+                  Diagnostics.set t.diagnostics (`Dune (id, uri, lsp_of_dune d)));
+            Diagnostics.send t.diagnostics
+          in
+          let handler = Client.Handler.create ~diagnostic () in
+          Fiber.return ((), init, Some handler)))
+
+let run t : (unit, state) result Fiber.t =
+  let open Fiber.O in
+  match Bin.which "dune" with
+  | None -> Fiber.return (Error Binary_not_found)
+  | Some bin -> (
+    let bin = Fpath.to_string bin in
+    (* TODO this is wrong on windows *)
+    let dev_null () = Unix.openfile "/dev/null" [ Unix.O_RDWR ] 0o666 in
+    let stdin = dev_null () in
+    let stdout = dev_null () in
+    let stderr = dev_null () in
+    let pid =
+      let args = Array.of_list [ bin; "rpc"; "--help=plain" ] in
+      Unix.create_process bin args stdin stdout stderr |> Stdune.Pid.of_int
+    in
+    let* status = Scheduler.wait_for_process t.scheduler pid in
+    match status with
+    | Unix.WEXITED 0 ->
+      let+ () = run_rpc t bin in
+      Ok ()
+    | _ -> Fiber.return (Error Out_of_date))
```

Comment:
If dune is not found in Path, the error will be a bit cryptic for users here

Reply:
Yes, that's why this function checks if the binary exists first :)